tools/pyUANamespace/ua_namespace.py

```diff
@@ -527,86 +529,38 @@ def printDotGraphWalk(self, depth=1, filename="out.dot", rootNode=None, followIn
-
-    logger.debug("Building connectivity matrix for node order optimization.")
-    # Set column 0 to contain the node
-    for node in self.nodes:
-      nind = self.nodes.index(node)
-      nmatrix[nind][0] = node
-
-    # Determine the dependencies of all nodes
-    for node in self.nodes:
-      nind = self.nodes.index(node)
-      #print "Examining node " + str(nind) + " " + str(node)
-      for ref in node.getReferences():
-        if isinstance(ref.target(), opcua_node_t):
-          tind = self.nodes.index(ref.target())
-          # Typedefinition of this node has precedence over this node
-          if ref.referenceType() in typeRefs and ref.isForward():
-            nmatrix[nind][tind+1] += 1
-          # isSubTypeOf/typeDefinition of this node has precedence over this node
-          elif ref.referenceType() in subTypeRefs and not ref.isForward():
-            nmatrix[nind][tind+1] += 1
-          # Else the target depends on us
-          elif ref.isForward():
-            nmatrix[tind][nind+1] += 1
-
-    logger.debug("Using Djikstra topological sorting to determine printing order.")
-    reorder = []
-    while len(reorder) < len(self.nodes):
-      (nind, node, w) = self.__reorder_getMinWeightNode__(nmatrix)
-      #print  str(100*float(len(reorder))/len(self.nodes)) + "% " + str(w) + " " + str(node) + " " + str(node.browseName())
-      reorder.append(node)
-      for ref in node.getReferences():
-        if isinstance(ref.target(), opcua_node_t):
-          tind = self.nodes.index(ref.target())
-          if ref.referenceType() in typeRefs and ref.isForward():
-            nmatrix[nind][tind+1] -= 1
-          elif ref.referenceType() in subTypeRefs and not ref.isForward():
-            nmatrix[nind][tind+1] -= 1
-          elif ref.isForward():
-            nmatrix[tind][nind+1] -= 1
-      nmatrix[nind][0] = None
-    self.nodes = reorder
-    logger.debug("Nodes reordered.")
-    return
+    #Kahn'S algorithm
+    #https://algocoding.wordpress.com/2015/04/05/topological-sorting-python/
+    relevant_references = ["HasSubtype", "HasTypeDefinition"]
+    in_degree = { u : 0 for u in self.nodes }     # determine in-degree
+    for u in self.nodes:                          # of each node
+      for ref in u.getReferences():
+       if isinstance(ref.target(), opcua_node_t):
+         if ref.referenceType().browseName() in relevant_references and ref.isForward():
+           in_degree[ref.target()] += 1
+    
+    Q = deque()                 # collect nodes with zero in-degree
+    for u in in_degree:
+      if in_degree[u] == 0:
+        Q.appendleft(u)
+ 
+    L = []     # list for order of nodes
+    
+    while Q:
+      u = Q.pop()          # choose node of zero in-degree
+      L.append(u)          # and 'remove' it from graph
+      for ref in u.getReferences():
+       if isinstance(ref.target(), opcua_node_t):
+         if ref.referenceType().browseName() in relevant_references and ref.isForward():
```

Comment:
The previous code took also the subtypes of the relevant references into account. Maybe, we should add this here too